x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java

```diff
@@ -256,7 +293,69 @@ protected void masterOperation(
             listener::onFailure
         );
 
-        checkModelIdAgainstTags(config.getModelId(), modelIdTagCheckListener);
+        if (config.isPackagedModel()) {
+            resolvePackageConfig(config.getModelId(), ActionListener.wrap(resolvedModelPackageConfig -> {
+                try {
+                    trainedModelConfig.validateNoPackageOverrides();
+                } catch (ValidationException e) {
+                    listener.onFailure(e);
+                }
+                modelPackageConfigHolder.set(resolvedModelPackageConfig);
+                setTrainedModelConfigFieldsFromPackagedModel(trainedModelConfig, resolvedModelPackageConfig);
+
+                checkModelIdAgainstTags(trainedModelConfig.getModelId(), modelIdTagCheckListener);
+            }, listener::onFailure));
+        } else {
+            checkModelIdAgainstTags(config.getModelId(), modelIdTagCheckListener);
+        }
+    }
+
+    private void setTrainedModelConfigFieldsFromPackagedModel(
+        TrainedModelConfig.Builder trainedModelConfig,
+        ModelPackageConfig resolvedModelPackageConfig
+    ) throws IOException {
+        trainedModelConfig.setDescription(resolvedModelPackageConfig.getDescription());
+        trainedModelConfig.setModelType(TrainedModelType.fromString(resolvedModelPackageConfig.getModelType()));
+        trainedModelConfig.setMetadata(resolvedModelPackageConfig.getMetadata());
+        trainedModelConfig.setInferenceConfig(
+            parseInferenceConfigFromModelPackage(
+                resolvedModelPackageConfig.getInferenceConfigSource(),
+                xContentRegistry,
+                LoggingDeprecationHandler.INSTANCE
+            )
+        );
+        trainedModelConfig.setTags(resolvedModelPackageConfig.getTags());
+        trainedModelConfig.setModelPackageConfig(
+            new ModelPackageConfig.Builder(resolvedModelPackageConfig).resetPackageOnlyFields().build()
+        );
+    }
+
+    private void triggerModelFetchIfNecessary(
+        String modelId,
+        ModelPackageConfig modelPackageConfig,
+        boolean waitForCompletion,
```

Comment:
`waitForCompletion` still misses some parts in the package loader code. I will do them in a separate PR.